lsp/src/main/kotlin/org/gradle/declarative/lsp/DeclarativeTextDocumentService.kt

```diff
@@ -502,54 +561,35 @@ private class ValueFactoryIndexStore() {
     }
 }
 
-private class ValueFactoryIndex(storeEntry: DocumentStoreEntry) {
+// Extension functions -------------------------------------------------------------------------------------------------
 
-    private val index: Map<FqName, List<LabelAndInsertText>> by lazy {
-        val schema = storeEntry.unionSchema
-        indexValueFactories(schema, schema.topLevelReceiverType, "")
+private fun DataTypeRef.toSimpleName(genericTypeSubstitution: Map<DataType.TypeVariableUsage, DataType> = emptyMap()): String =
+    when (this) {
+        is DataTypeRef.Name -> fqName.simpleName
+        is DataTypeRef.NameWithArgs -> "${fqName.simpleName}<${typeArguments.joinToString { it.toSimpleName(genericTypeSubstitution) }}>"
+        is DataTypeRef.Type -> dataType.toString()
     }
 
-    // TODO: currently this index contains all value factories from the whole schema
-    //  going forward we should make it take into account which value factory is available
-    //  in which block, because only the ones defined at the top level are available everywhere
-
-    fun factoriesForProperty(fqName: FqName): List<LabelAndInsertText>? = index[fqName]
-
-    private fun indexValueFactories(analysisSchema: AnalysisSchema, type: DataClass, namePrefix: String): Map<FqName, List<LabelAndInsertText>> {
-        val factoryIndex = mutableMapOf<FqName, List<LabelAndInsertText>>()
-
-        type.memberFunctions
-            .filter { it.semantics is FunctionSemantics.Pure && it.returnValueType is DataTypeRef.Name }
-            .forEach {
-                val indexKey = (it.returnValueType as DataTypeRef.Name).fqName
-                val labelAndInsertText = LabelAndInsertText("$namePrefix${computeCompletionLabel(it)}", "$namePrefix${computeCompletionInsertText(it, analysisSchema)}")
-                factoryIndex.merge(indexKey, listOf(labelAndInsertText)) { oldVal, newVal -> oldVal + newVal }
-            }
-
-        val typeRefContext = SchemaTypeRefContext(analysisSchema)
-        type.properties.forEach {
-            when (val propType = typeRefContext.resolveRef(it.valueType)) {
-                is DataClass -> {
-                    val propName = it.name
-                    val propIndex = indexValueFactories(analysisSchema, propType, "$namePrefix${propName}.")
-                    propIndex.forEach { (key, value) -> factoryIndex.merge(key, value) { oldVal, newVal -> oldVal + newVal } }
-                }
-                else -> Unit
+private fun DataType.ParameterizedTypeInstance.TypeArgument.toSimpleName(
+    genericTypeSubstitution: Map<DataType.TypeVariableUsage, DataType> = emptyMap()
+) =
+    when (this) {
+        is DataType.ParameterizedTypeInstance.TypeArgument.ConcreteTypeArgument ->
```

Comment:
💭 Does this properly handle the type substitution cases with the type variable usage appearing not directly in the type's type arguments but deeper?

Consider this function: `fun <T> listOfListOfT(t: T): List<List<T>>`. Its return type is a parameterized type instance `List<List<T>>`. With a substitution of `T: Int`, it should be rendered as `List<List<Int>>`.

Reply:
Problem addressed, please take a look.